syne_tune/optimizer/schedulers/regularized_evolution.py

```diff
@@ -0,0 +1,132 @@
+        https://colab.research.google.com/github/google-research/google-research/blob/master/evolution/regularized_evolution_algorithm/regularized_evolution.ipynb
+
+
+        Parameters
+        ----------
+        config_space: dict
+            Configuration space for trial evaluation function
+        metric : str
+            Name of metric to optimize, key in result's obtained via
+            `on_trial_result`
+        mode : str
+            Mode to use for the metric given, can be 'min' or 'max', default to 'min'.
+        population_size : int
+            Size of the population.
+        sample_size : int
+            Size of the candidate set to obtain a parent for the mutation.
+        random_seed : int
+            Seed for the random number generation. If set to None, use random seed.
+        """
+        super(RegularizedEvolution, self).__init__(config_space=config_space)
+        self.metric = metric
+        self.mode = mode
+        self.population_size = population_size
+        self.sample_size = sample_size
+        self.population = deque()
+        if random_seed is None:
+            random_seed = np.random.randint(0, 2 ** 32)
+        self._random_state = np.random.RandomState(random_seed)
+
+    def mutate_arch(self, config):
+
+        child_config = copy.deepcopy(config)
+
+        # pick random hyperparameter and mutate it
+        hypers = []
+        for k, v in self.config_space.items():
+            if isinstance(v, Domain):
+                hypers.append(k)
+        name = self._random_state.choice(hypers)
+        hyperparameter = self.config_space[name]
+
+        if isinstance(hyperparameter, Categorical):
+            # drop current values from potential choices to not sample the same value again
+            choices = copy.deepcopy(hyperparameter.categories)
+            choices.remove(config[name])
+
+            new_value = self._random_state.choice(choices)
+
+        elif isinstance(hyperparameter, Float):
+            new_value = Float.sample(random_state=self._random_state)
+
+        elif isinstance(hyperparameter, Integer):
+            new_value = Integer.sample(random_state=self._random_state)
+
+        child_config[name] = new_value
+
+        return child_config
+
+    def sample_random_config(self):
+        return {k: v.sample(random_state=self._random_state) if isinstance(v, Domain) else v for k, v in
+                self.config_space.items()}
+
+    def _suggest(self, trial_id: int) -> Optional[TrialSuggestion]:
+
+        if len(self.population) < self.population_size:
+            config = self.sample_random_config()
+        else:
+            candidates = self._random_state.choice(list(self.population), size=self.sample_size)
+            parent = max(candidates, key=lambda i: i.score)
+
+            config = self.mutate_arch(parent.config)
+
+        return TrialSuggestion.start_suggestion(config=config)
```

Comment:
In the original paper they used it for NAS without weight sharing. I am not sure how that would work with checkpointing?

Reply:
To be clear, it is just a point of curiosity and not something that we would necessarily need for this PR.

We could support an option that also resume from the checkpoint of the mutated trial to save compute time (as in PBT). Of-course, it is only possible for some config-space (learning-rate or batch-size for instance not number of layers or cells).

```suggestion
        if self.resume_from_checkpoint:
          return TrialSuggestion.start_suggestion(config=config, checkpoint_trial_id=config_trial_id)
        else:
          return TrialSuggestion.start_suggestion(config=config)
```

Reply:
Another reason this should be a searcher. Then you can use it with PBT scheduling as well.

Reply:
I do not think it is equivalent, using REA as a searcher in PBT gives a different algorithms than just adding checkpointing support in REA.

In any case, I think Aaron should decide what is the best approach since he contributed the method. I think they are trade-offs. On one hand we can reuse REA as a searcher in other cases if implemented as a searcher, but it also makes the method example more complex to grasp (a few base classes and edge case to be taken care of).

Reply:
It's a tricky case. I will make REA a searcher for consistency. However, it's not a perfect solution. For example it might be confusing to users to be able to pass points_to_evaluate since the algorithm always starts from a random population. We might have to rethink our interfaces. I also agree with David, that REA combined with PBT is a new algorithm compared to just adding checkpoint support.

Reply:
Thanks, and as I said, I can help.
We should eat our own dog food, the BaseSearcher API was commonly agreed upon, it does not help if only I make use of it.

Reply:
points_to_evaluate is taken from Ray Tune, it just makes our code more modular, because you can specify the first config(s) to try, which I think can make sense for any scheduler.

Reply:
I stop here, but there are shades of "difficult to grasp". You can do anything from scratch every time you code a scheduler, and then maybe the single isolated code may be easier to grasp, because you never need to learn about common concepts. But the overall codebase will be a mess of duplicated code, and things like controlling the random seed or using user-specified default configs will be very hard.

As we otherwise put a big emphasis on clean code structure, we should follow it here as well.